### PR TITLE
fix: merged schemas reference resolving

### LINKED ISSUE #540: Cannot find reference with allOf
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Fastify version

4.5.3

### Plugin version

5.3.0

### Node.js version

16

### Operating system

macOS

### Operating system version (i.e. 20.04, 11.3, 10)

12.2

### Description

I suspect it comes from https://github.com/fastify/fastify/pull/4049 shipped with fastify 4.2.0 but I cannot figure what is wrong with my case.
It used to work with  fastify 4.1.0 .

Maybe you have a clue.

### Steps to Reproduce

Schema is :
```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "Test",
  "definitions": {
    "Problem": {
      "type": "object",
      "required": [
        "type"
      ],
      "additionalProperties": true,
      "properties": {
        "type": {
          "type": "string"
        }
      }
    },
    "ValidationFragment": {
      "type": "string"
    },
    "ValidationErrorProblem": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/Problem"
        },
        {
          "type": "object",
          "properties": {
            "validation": {
              "$ref": "#/definitions/ValidationFragment"
            }
          }
        }
      ]
    }
  },
  "additionalProperties": false,
  "type": "object"
}
```
Code is
```ts
import fastify from 'fastify';

const server = fastify();

const schema = require('./schema.json');
server.addSchema(schema);

server.post<{
}>(
  '/dummy',
  {
    schema: {
      response: {
        '400': {$ref: 'Test#/definitions/ValidationErrorProblem'},
      },
    }
  },
  async (_request, _reply) => {
    return 'OK'
  });

server.ready();
```
Error is :
```
node_modules/fastify/lib/route.js:326
              throw new FST_ERR_SCH_SERIALIZATION_BUILD(opts.method, url, error.message)
                    ^
FastifyError [Error]: Failed building the serialization schema for POST: /dummy, due to error Cannot find reference "#/definitions/ValidationFragment"
```

### Expected Behavior

It used to work with fastify 4.1.0. I would expect FJS to compile the schema.

### PR DESCRIPTION
**Pull request by ivan-tymoshenko**
Fix #540

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -40,12 +40,47 @@ function isValidSchema (schema, name) {
   }
 }
 
-function mergeLocation (location, key) {
-  return {
-    schema: location.schema[key],
-    schemaId: location.schemaId,
-    jsonPointer: location.jsonPointer + '/' + key,
-    isValidated: location.isValidated
+class Location {
```
Comment (Eomm) [inline index.js:43]: Did you evaluate the `URL` class?

Comment (ivan-tymoshenko) [reply index.js:43]: Sorry, I didn't get you.

File: test/allof.test.js [hunk 2]
```diff
@@ -437,3 +437,48 @@ test('object with external $refs in allOf', (t) => {
...
+          type: 'object',
+          allOf: [
+            {
+              $ref: '#/definitions/Problem'
+            },
+            {
+              type: 'object',
+              properties: {
+                validation: {
+                  $ref: '#/definitions/ValidationFragment'
```
Comment (Eomm) [inline test/allof.test.js:469]: what about an absolute ref like `schema:app:foo#/definitions/hello`?

File: index.js (hunk 1 above)
Comment (Eomm) [reply index.js:43]: The [URL](https://nodejs.org/api/url.html#class-url) class has many convenient methods to calculate if the URL is absolute and relative and compose it.

Looking at this `Location` class, I have afraid that we are going to replicate the `URL` class in future.

So I wanted to point it out to reduce the burden you could face by developing it

File: test/allof.test.js (hunk 2 above)
Comment (ivan-tymoshenko) [reply test/allof.test.js:469]: Yes, they are working. This PR doesn't change anything for absolute refs. The problem was that when we create merged schema we should have two locations to resolve local refs:
1. We need location inside new "merged" schema in case if we have anyof/oneof .., inside merged schema.
2. We original location to be able to resolve local reference in this schema, but outside of merged schema.

It's hard to explain simply) #480 Fixed the first case, but broke the second. This PR should support both.

**Comment by ivan-tymoshenko** [comment]
@Eomm I can't replace the hole Location class with an URL, we've got some custom logic here. I can use URL instead of split/join for schemaId and jsonPointer, but I'm not sure that it's a good idea. All operation that we've got pretty simple and URL is super slow in compare with them.
I can't replace the hole Location class with an URL, we've got some custom logic here. I can use URL instead of split/join for schemaId and jsonPointer, but I'm not sure that it's a good idea. All operation that we've got pretty simple and URL is super slow in compare with them.

File: index.js [hunk 3]
```diff
@@ -40,12 +40,47 @@ function isValidSchema (schema, name) {
...
+  constructor (schema, schemaId, jsonPointer = '#', isValidated = false) {
+    this.schema = schema
+    this.schemaId = schemaId
+    this.jsonPointer = jsonPointer
+    this.isValidated = isValidated
+
+    this.isMerged = false
+    this.mergedSchemaId = null
+    this.mergedJsonPointer = null
```
Comment (climba03003) [inline index.js:52]: It is currently hard to follow why does it fix the issue and how it is fixed.
Would you mind to add some comment of step flow why it works? and how the usage of `mergedJsonPointer`?

File: index.js (hunk 1 above)
Comment (climba03003) [reply index.js:43]: I don't think they are the same and `URL` is really slow.
The usage here are only `string` concatenation.

File: index.js [hunk 4]
```diff
@@ -40,12 +40,47 @@ function isValidSchema (schema, name) {
...
+      return this.mergedSchemaId + this.mergedJsonPointer
+    }
+    return this.schemaId + this.jsonPointer
+  }
+
+  addMergedSchema (schemaId, jsonPointer = '#') {
+    this.isMerged = true
+    this.mergedSchemaId = schemaId
+    this.mergedJsonPointer = jsonPointer
   }
```
Comment (mcollina) [inline index.js:84]: Can you move this class into its own file under lib?

File: index.js (hunk 3 above)
Comment (ivan-tymoshenko) [reply index.js:52]: @climba03003 @Eomm  I will explain with an example. Sorry, it will not be simple because the case is not simple.

```js
{
  $id: 'root',
  type: 'object',
  properties: {
    property1: { type: 'string' },
    property2: {
      type: 'object',
      allOf: [
        {
          properties: {
            property21: { $ref: '#/properties/property1' }
          }
        },
        {
          properties: {
            property22: {
              anyOf: [
                { type: 'string' },
                { type: 'number' }
              ]
            }
          }
        }
      ]
    }
  }
}
```

1. When we meet property `allOf`, we create a merged schema and create a new unique id for it. 
https://github.com/fastify/fast-json-stringify/blob/e6b3c5fb6e1dc47f1c71560adf2ef3363bbf2b70/index.js#L458
If we meet anyOf/oneOf/if inside the merged schema, we create a reference using the merged schema id.
`property22` will have the next reference: `merged_12...223#/properties/property22`. To create this reference, we should know the merged schema id.

2. If somewhere inside the `allOf` schemas we meet a local reference like  `'#/properties/property1'` we have to resolve it. To do this, we should know the original schema reference, which is `root` in this case.

So that is why we need to keep two schemas ids when we build a serializer inside the `allOf` property.

Comment (ivan-tymoshenko) [reply index.js:52]: The `Location` class now has two methods for getting schema id.
1. `getSchemaId` - you should use it when you want to CREATE a new reference to some place in the schema. For example, for validation.
2. `getOriginSchemaId` - you should use it when you want to RESOLVE already existing references.

I added corresponding comments in the code.

**Review by climba03003 — APPROVED** [review]
LGTM. 
Thanks for the explanation, it's now more clear to me.

**Comment by remidewitte** [comment]
Thanks. I will test asap :)
